spec/formatting.md

```diff
@@ -284,9 +488,48 @@ when *   {Other match}
 
 4. The pattern `{Exact match}` of the most preferred `1` variant is selected.
 
-## Handling Bidirectional Text
+## Formatting
+
+After _pattern selection_,
+each _text_ and _expression_ of the selected _pattern_ must be resolved and formatted.
+
+_Formatting_ is a mostly implementation-defined process,
+as it depends on the implementation's shape for resolved values
+and the target shape of the formatting.
+
+Formatting errors MAY be emitted during _formatting_,
+as the successful _resolution_ of an _expression_ does not guarantee success in its _formatting_.
+A formatter MAY provide a value to use in such a case instead of a _fallback value_.
+
+_Formatting_ MAY produce formatted messages in the following shapes,
+as well as any others:
+
+- A single concatenated string.
+- A flat sequence of objects corresponding to each resolved value.
+- A hierarchical structure of objects that group spans of resolved values,
+  such as sequences delimited by "open" and "close" _function_ _annotations_.
+
+Implementations SHOULD provide _formatting_ output shapes that match user needs,
+such that e.g. re-parsing a message formatted as a string is not required.
```

Comment:
Not sure I understand this -- should it be "re-parsing an output shape that is a single formatted string"?

Reply:
I think so? As in, there should be no need for someone to format MF2 to a string, and then right away try to parse that string.

Reply:
I understand now, but I still find the text as it is a bit hard to follow. Maybe it would be easier to read split into two sentences, like:

```
Implementations SHOULD provide _formatting_ output shapes that match user needs.
For example, a user should never need to immediately re-parse the string output by the formatter.
Instead, the formatter itself should directly produce a more structured representation.
```

That's still a bit verbose, but the point to get across is that `string` shouldn't serve as an intermediate data structure, I think.

Reply:
Would this be clearer?
```suggestion
including situations that require further processing of formatted messages.
```

Reply:
Applied the above change.